File: locationkit-java-sample/app/src/main/java/com/huawei/locationsample6/location/fusedlocation/RequestLocationUpdatesWithCallbackActivity.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2021-2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.huawei.locationsample6.location.fusedlocation;

import java.util.List;

import android.location.Location;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;

import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hmf.tasks.Task;
import com.huawei.hms.location.FusedLocationProviderClient;
import com.huawei.hms.location.LocationAvailability;
import com.huawei.hms.location.LocationCallback;
import com.huawei.hms.location.LocationRequest;
import com.huawei.hms.location.LocationResult;
import com.huawei.hms.location.LocationServices;
import com.huawei.hms.location.LocationSettingsRequest;
import com.huawei.hms.location.LocationSettingsResponse;
import com.huawei.hms.location.SettingsClient;
import com.huawei.locationsample6.LogInfoUtil;
import com.huawei.locationsample6.R;
import com.huawei.locationsample6.util.NotificationUtil;
import com.huawei.logger.LocationLog;

/**
 * Example of Using requestLocationUpdates and removeLocationUpdates.
 * Requests a location update and calls back on the specified Looper thread. This method requires that the requester
 * process exist for continuous callback.
 * If you still want to receive the callback after the process is killed, see requestLocationUpdates (LocationRequest
 * request,PendingIntent callbackIntent)
 *
 * @since 2020-5-11
 */
public class RequestLocationUpdatesWithCallbackActivity extends LocationBaseActivity implements OnClickListener {
    public static final String TAG = "LocationUpdatesCallback";

    LocationCallback mLocationCallback;

    LocationRequest mLocationRequest;

    private FusedLocationProviderClient mFusedLocationProviderClient;

    private SettingsClient mSettingsClient;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_location_request_location_updates_callback);
        // Button click listeners
        findViewById(R.id.location_requestLocationUpdatesWithCallback).setOnClickListener(this);
        findViewById(R.id.location_removeLocationUpdatesWithCallback).setOnClickListener(this);
        findViewById(R.id.location_enableBackgroundLocation).setOnClickListener(this);
        findViewById(R.id.location_disableBackgroundLocation).setOnClickListener(this);
        addLogFragment();
        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this);
        mSettingsClient = LocationServices.getSettingsClient(this);
        mLocationRequest = new LocationRequest();
        // Sets the interval for location update (unit: Millisecond)
        mLocationRequest.setInterval(5000);
        // Sets the priority
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
        // Sets the type of the returned coordinate:
        // COORDINATE_TYPE_WGS84 Indicates that the 84 coordinate is returned.
        // COORDINATE_TYPE_GCJ02 Indicates that the 02 coordinate is returned. The default value is COORDENATE_TYPE_WGS84.
        mLocationRequest.setCoordinateType(LocationRequest. COORDINATE_TYPE_WGS84);
        if (null == mLocationCallback) {
            mLocationCallback = new LocationCallback() {
                @Override
                public void onLocationResult(LocationResult locationResult) {
                    if (locationResult != null) {
                        List<Location> locations = locationResult.getLocations();
                        if (!locations.isEmpty()) {
                            for (Location location : locations) {
                                LocationLog.i(TAG,
                                    "onLocationResult location[Longitude,Latitude,Accuracy]:" + location.getLongitude()
                                        + "," + location.getLatitude() + "," + location.getAccuracy());
                            }
                        }
                    }
                }

                @Override
                public void onLocationAvailability(LocationAvailability locationAvailability) {
                    if (locationAvailability != null) {
                        boolean flag = locationAvailability.isLocationAvailable();
                        LocationLog.i(TAG, "onLocationAvailability isLocationAvailable:" + flag);
                    }
                }
            };
        }
    }

    /**
     * Requests a location update and calls back on the specified Looper thread.
     */
    private void requestLocationUpdatesWithCallback() {
        Log.i(TAG, "requestLocationUpdatesWithCallback");
        LogInfoUtil.getLogInfo(this);
        try {
            LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder();
            builder.addLocationRequest(mLocationRequest);
            LocationSettingsRequest locationSettingsRequest = builder.build();
            // Before requesting location update, invoke checkLocationSettings to check device settings.
            Task<LocationSettingsResponse> locationSettingsResponseTask =
                mSettingsClient.checkLocationSettings(locationSettingsRequest);
            locationSettingsResponseTask.addOnSuccessListener(locationSettingsResponse -> {
                Log.i(TAG, "check location settings success");
                mFusedLocationProviderClient
                    .requestLocationUpdates(mLocationRequest, mLocationCallback, Looper.getMainLooper())
                    .addOnSuccessListener(aVoid -> {
                        LocationLog.i(TAG, "requestLocationUpdatesWithCallback onSuccess");
                    })
                    .addOnFailureListener(e -> {
                        LocationLog.e(TAG, "requestLocationUpdatesWithCallback onFailure:" + e.getMessage());
                    });
            }).addOnFailureListener(e -> LocationLog.e(TAG, "checkLocationSetting onFailure:" + e.getMessage()));
        } catch (Exception e) {
            LocationLog.e(TAG, "requestLocationUpdatesWithCallback exception:" + e.getMessage());
        }
    }

    @Override
    protected void onDestroy() {
        // Removed when the location update is no longer required.
        removeLocationUpdatesWithCallback();
        disableBackgroundLocation();
        super.onDestroy();
    }

    /**
     * Removed when the location update is no longer required.
     */
    private void removeLocationUpdatesWithCallback() {
        try {
            Task<Void> voidTask = mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
            voidTask.addOnSuccessListener(new OnSuccessListener<Void>() {
                @Override
                public void onSuccess(Void aVoid) {
                    LocationLog.i(TAG, "removeLocationUpdatesWithCallback onSuccess");
                }
            }).addOnFailureListener(new OnFailureListener() {
                @Override
                public void onFailure(Exception e) {
                    LocationLog.e(TAG, "removeLocationUpdatesWithCallback onFailure:" + e.getMessage());
                }
            });
        } catch (Exception e) {
            LocationLog.e(TAG, "removeLocationUpdatesWithCallback exception:" + e.getMessage());
        }
    }

    private void enableBackgroundLocation() {
        mFusedLocationProviderClient
                .enableBackgroundLocation(NotificationUtil.NOTIFICATION_ID, NotificationUtil.getNotification(this))
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        LocationLog.i(TAG, "enableBackgroundLocation onSuccess");
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        LocationLog.e(TAG, "enableBackgroundLocation onFailure:" + e.getMessage());
                    }
                });
    }

    private void disableBackgroundLocation() {
        mFusedLocationProviderClient.disableBackgroundLocation()
                .addOnSuccessListener(new OnSuccessListener<Void>() {
                    @Override
                    public void onSuccess(Void aVoid) {
                        LocationLog.i(TAG, "disableBackgroundLocation onSuccess");
                    }
                })
                .addOnFailureListener(new OnFailureListener() {
                    @Override
                    public void onFailure(Exception e) {
                        LocationLog.e(TAG, "disableBackgroundLocation onFailure:" + e.getMessage());
                    }
                });
    }

    @Override
    public void onClick(View v) {
        try {
            switch (v.getId()) {
                case R.id.location_requestLocationUpdatesWithCallback:
                    requestLocationUpdatesWithCallback();
                    break;
                case R.id.location_removeLocationUpdatesWithCallback:
                    removeLocationUpdatesWithCallback();
                    break;
                case R.id.location_enableBackgroundLocation:
                    enableBackgroundLocation();
                    break;
                case R.id.location_disableBackgroundLocation:
                    disableBackgroundLocation();
                    break;
                default:
                    break;
            }
        } catch (Exception e) {
            Log.e(TAG, "RequestLocationUpdatesWithCallbackActivity Exception:" + e);
        }
    }
}

File: locationkit-kotlin-sample/app/src/main/java/com/hms/locationsample6/fusedlocation/RequestLocationUpdatesWithCallbackActivity.kt
/**
 * Copyright 2020 Huawei Technologies co, Ltd All
 * Rights reserved
 * Licenced under the Apache License,Version 2.0(the "License");
 * you may not use this file except in compliance with license
 * you may obtain a copy of the license at
 * http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by application law or agreed to in writing software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permission and
 * limitations under the License
 */

package com.hms.locationsample6.fusedlocation

import android.Manifest.permission.ACCESS_COARSE_LOCATION
import android.Manifest.permission.ACCESS_FINE_LOCATION
import android.content.pm.PackageManager.PERMISSION_GRANTED
import android.location.Location
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.util.Log
import android.view.View
import androidx.core.app.ActivityCompat.checkSelfPermission
import androidx.core.app.ActivityCompat.requestPermissions
import com.hms.locationsample6.R
import com.hms.locationsample6.activity.BaseActivity
import com.hms.locationsample6.logger.LocationLog
import com.hms.locationsample6.utils.NotificationUtil
import com.huawei.hmf.tasks.Task
import com.huawei.hms.location.*
import kotlinx.android.synthetic.main.activity_request_location_updates_callback.*
import kotlinx.android.synthetic.main.item_background_location.*

class RequestLocationUpdatesWithCallbackActivity : BaseActivity(), View.OnClickListener {
    companion object {
        private const val TAG = "LocationUpdatesCallback"
    }

    // the callback of the request
    private var mLocationCallback: LocationCallback? = null

    private var mLocationRequest: LocationRequest? = null

    private lateinit var fusedLocationProviderClient: FusedLocationProviderClient

    private lateinit var settingsClient: SettingsClient

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_request_location_updates_callback)
        location_requestLocationUpdatesWithCallback.setOnClickListener(this)
        location_removeLocationUpdatesWithCallback.setOnClickListener(this)
        location_enableBackgroundLocation.setOnClickListener(this)
        location_disableBackgroundLocation.setOnClickListener(this)
        addLogFragment()
        // create fusedLocationProviderClient
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)
        // create settingsClient
        settingsClient = LocationServices.getSettingsClient(this)
        mLocationRequest = LocationRequest().apply {
            interval = 1000
            needAddress = true
            priority = LocationRequest.PRIORITY_HIGH_ACCURACY
            // Sets the type of the returned coordinate:
            // COORDINATE_TYPE_WGS84 Indicates that the 84 coordinate is returned.
            // COORDINATE_TYPE_GCJ02 Indicates that the 02 coordinate is returned. The default value is COORDENATE_TYPE_WGS84.
            coordinateType = LocationRequest.COORDINATE_TYPE_WGS84
        }
        // Set the interval for location updates, in milliseconds.
        // set the priority of the request
        if (null == mLocationCallback) {
            mLocationCallback = object : LocationCallback() {
                override fun onLocationResult(locationResult: LocationResult?) {
                    if (locationResult != null) {
                        val locations: List<Location> =
                            locationResult.getLocations()
                        if (locations.isNotEmpty()) {
                            for (location in locations) {
                                LocationLog.i(
                                    TAG,
                                    "onLocationResult location[Longitude,Latitude,Accuracy]:${location.longitude} , ${location.latitude} , ${location.accuracy}"
                                )
                            }
                        }
                    }
                }

                override fun onLocationAvailability(locationAvailability: LocationAvailability?) {
                    locationAvailability?.let {
                        val flag: Boolean = locationAvailability.isLocationAvailable
                        LocationLog.i(TAG, "onLocationAvailability isLocationAvailable:$flag")
                    }
                }
            }
        }

        // check location permission
        if (Build.VERSION.SDK_INT <= Build.VERSION_CODES.P) {
            Log.i(TAG, "sdk < 28 Q")
            if (checkSelfPermission(
                    this,
                    ACCESS_FINE_LOCATION
                ) != PERMISSION_GRANTED
                && checkSelfPermission(
                    this,
                    ACCESS_COARSE_LOCATION
                ) != PERMISSION_GRANTED
            ) {
                val strings = arrayOf(
                    ACCESS_FINE_LOCATION,
                    ACCESS_COARSE_LOCATION
                )
                requestPermissions(this, strings, 1)
            }
        } else {
            if (checkSelfPermission(
                    this@RequestLocationUpdatesWithCallbackActivity,
                    ACCESS_FINE_LOCATION
                ) != PERMISSION_GRANTED && checkSelfPermission(
                    this@RequestLocationUpdatesWithCallbackActivity,
                    ACCESS_COARSE_LOCATION
                ) != PERMISSION_GRANTED && checkSelfPermission(
                    this@RequestLocationUpdatesWithCallbackActivity,
                    "android.permission.ACCESS_BACKGROUND_LOCATION"
                ) != PERMISSION_GRANTED
            ) {
                val strings = arrayOf(
                    ACCESS_FINE_LOCATION,
                    ACCESS_COARSE_LOCATION,
                    "android.permission.ACCESS_BACKGROUND_LOCATION"
                )
                requestPermissions(this, strings, 2)
            }
        }
    }

    /**
     * function：Requests location updates with a callback on the specified Looper thread.
     * first：use SettingsClient object to call checkLocationSettings(LocationSettingsRequest locationSettingsRequest) method to check device settings.
     * second： use  FusedLocationProviderClient object to call requestLocationUpdates (LocationRequest request, LocationCallback callback, Looper looper) method.
     */
    private fun requestLocationUpdatesWithCallback() {
        try {
            val builder = LocationSettingsRequest.Builder()
            builder.addLocationRequest(mLocationRequest)
            val locationSettingsRequest = builder.build()
            // check devices settings before request location updates.
            //Before requesting location update, invoke checkLocationSettings to check device settings.
            val locationSettingsResponseTask: Task<LocationSettingsResponse> =
                settingsClient.checkLocationSettings(locationSettingsRequest)

            locationSettingsResponseTask.addOnSuccessListener { locationSettingsResponse: LocationSettingsResponse? ->
                Log.i(TAG, "check location settings success  {$locationSettingsResponse}")
                // request location updates
                fusedLocationProviderClient.requestLocationUpdates(
                    mLocationRequest,
                    mLocationCallback,
                    Looper.getMainLooper()
                )
                    .addOnSuccessListener {
                        LocationLog.i(TAG, "requestLocationUpdatesWithCallback onSuccess")
                    }
                    .addOnFailureListener { e ->
                        LocationLog.e(
                            TAG,
                            "requestLocationUpdatesWithCallback onFailure:${e.message}"
                        )
                    }
            }
                .addOnFailureListener { e: Exception ->
                    LocationLog.e(TAG, "checkLocationSetting onFailure:${e.message}")
                }
        } catch (e: Exception) {
            LocationLog.e(TAG, "requestLocationUpdatesWithCallback exception:${e.message}")
        }
    }

    override fun onDestroy() {
        // don't need to receive callback
        removeLocationUpdatesWithCallback()
        disableBackgroundLocation()
        super.onDestroy()
    }

    /**
     * remove the request with callback
     */
    private fun removeLocationUpdatesWithCallback() {
        try {
            fusedLocationProviderClient.removeLocationUpdates(mLocationCallback)
                .addOnSuccessListener {
                    LocationLog.i(
                        TAG,
                        "removeLocationUpdatesWithCallback onSuccess"
                    )
                }
                .addOnFailureListener { e ->
                    LocationLog.e(
                        TAG,
                        "removeLocationUpdatesWithCallback onFailure:${e.message}"
                    )
                }
        } catch (e: Exception) {
            LocationLog.e(
                TAG,
                "removeLocationUpdatesWithCallback exception:${e.message}"
            )
        }
    }

    private fun enableBackgroundLocation() {
        fusedLocationProviderClient
            .enableBackgroundLocation(
                NotificationUtil.NOTIFICATION_ID,
                NotificationUtil.getNotification(this)
            )
            .addOnSuccessListener {
                LocationLog.i(
                    TAG,
                    "enableBackgroundLocation onSuccess"
                )
            }
            .addOnFailureListener { e ->
                LocationLog.e(
                    TAG,
                    "enableBackgroundLocation onFailure:${e.message}"
                )
            }
    }

    private fun disableBackgroundLocation() {
        fusedLocationProviderClient.disableBackgroundLocation()
            .addOnSuccessListener {
                LocationLog.i(
                    TAG,
                    "disableBackgroundLocation onSuccess"
                )
            }
            .addOnFailureListener { e ->
                LocationLog.e(
                    TAG,
                    "disableBackgroundLocation onFailure:${e.message}"
                )
            }
    }

    override fun onClick(v: View) {
        try {
            when (v.id) {
                R.id.location_requestLocationUpdatesWithCallback -> requestLocationUpdatesWithCallback()
                R.id.location_removeLocationUpdatesWithCallback -> removeLocationUpdatesWithCallback()
                R.id.location_enableBackgroundLocation -> enableBackgroundLocation()
                R.id.location_disableBackgroundLocation -> disableBackgroundLocation()
            }
        } catch (e: Exception) {
            Log.e(
                TAG,
                "RequestLocationUpdatesWithCallbackActivity Exception:$e"
            )
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 1) {
            if (grantResults.size > 1 && grantResults[0] == PERMISSION_GRANTED && grantResults[1] == PERMISSION_GRANTED
            ) {
                Log.i(
                    TAG, "onRequestPermissionsResult: apply LOCATION PERMISSION successful"
                )
            } else {
                Log.i(
                    TAG, "onRequestPermissionsResult: apply LOCATION PERMISSSION  failed"
                )
            }
        }
        if (requestCode == 2) {
            if (grantResults.size > 2 && grantResults[2] == PERMISSION_GRANTED && grantResults[0] == PERMISSION_GRANTED && grantResults[1] == PERMISSION_GRANTED
            ) {
                Log.i(
                    TAG, "onRequestPermissionsResult: apply ACCESS_BACKGROUND_LOCATION successful"
                )
            } else {
                Log.i(
                    TAG, "onRequestPermissionsResult: apply ACCESS_BACKGROUND_LOCATION  failed"
                )
            }
        }
    }


}
